# Hodan Mohamed Abdi SBA23416_Data Preperation_CA1

# Characterisation of the data set- aps_failure_set (1).csv

<font color="blue"> Introduction: 

My assignment i have been asked by Haulage company to analyse a dataset based on data collected from heavy Scania trucks in everyday usage. The system in focus is the Air Pressure system (APS) which generates pressurised air that are utilized in various functions in a truck, such as braking and gear changes.

*The dataset’s <font color="green"> positive class consists of component failures for a specific component of the APS system. text</font> 

*The <font color="red"> negative class consists of trucks with failures for components not related to the APS

*The data consists of a subset of all available data, selected by experts.

*The main aim of this analysis will help determine the investment strategy for the company in the upcoming year.

All data wrangling, analysis, and visualizations must be generated using python.

The companies CTO also requires that i must include rationalize all the decisions that you have made in the report.

Assignment Minimum Requirements: 

Step 1 of the assignment is to: use the dataset contained within the file “aps_failure_set.csv”, conduct the following analysis and report with my findings as following step below:

1.Characterisation of the data set: Which incldues the size; How many number of attributes; if there are or has/does not have missing values, Also listing the number of observations. and idenfying what these Characterisation mean.

I will be importing the dataset with Panda library to verify what are the specific's

Step 2:Application of Data preparation & Evaluation Methods: Include the following:Cleaning, renaming, & Exploratory Data Analysis (EDA) to get a better understanding of the data-set by summarizing its main characteristics and often plotting them visually.

I have come to understand that this step is very important especially when we arrive at modelling the data to apply Machine learning. Plotting in EDA consists of Histograms, Box plot, Scatter plots.

Step 3: Use Principal Component Analysis, In this section, i will be explore what is perhaps one of the most broadly used of unsupervised algorithms, principal component analysis (PCA). PCA is fundamentally a dimensionality reduction algorithm, but it can also be useful as a tool for visualization, for noise filtering,

Step 4:Final Part: Curse of Dimensionality, I will end the assignment with my finding toward my analysis.

# 1. Characterisation of the data set

In [1]:
import pandas as pd
%matplotlib inline 
import numpy as np 
import matplotlib.pyplot as plt 
import seaborn as sns
sns.set() 

In [2]:
aps_failure_df =pd.read_csv ("aps_failure_set (2).csv")

<font color="green"> Size of the dataset text</font> : In this section i have identified the size of the dataset using the follow code below: Dataset size: (6000 , 171)

In [3]:
size = aps_failure_df.shape
print("Dataset Size:", size)

Dataset Size: (60000, 171)


<font color="green"> Data Information text</font>: In this section, I have highlighted the current datafrane which includes the following, range index, columns, dtypes and memory.

In [4]:
aps_failure_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 60000 entries, 0 to 59999
Columns: 171 entries, class to eg_000
dtypes: int64(1), object(170)
memory usage: 78.3+ MB


To get a better understanding of the dataset, I would need to identify rhe Row's and Columns and displaying the dtype value counts. as you can cledarly see within the below that the has 6000 Rows and 171 Colums.

In [5]:
print("The data set has {} rows and {} columns".format(aps_failure_df.shape[0], aps_failure_df.shape[1]))
display(aps_failure_df.describe())
display(aps_failure_df.head())
print(aps_failure_df.dtypes.value_counts())

The data set has 60000 rows and 171 columns


aa_000
count  6.000000e+04
mean   5.933650e+04
std    1.454301e+05
min    0.000000e+00
25%    8.340000e+02
50%    3.077600e+04
75%    4.866800e+04
max    2.746564e+06

class  aa_000 ab_000      ac_000 ad_000 ae_000 af_000 ag_000 ag_001 ag_002  \
0   neg   76698     na  2130706438    280      0      0      0      0      0   
1   neg   33058     na           0     na      0      0      0      0      0   
2   neg   41040     na         228    100      0      0      0      0      0   
3   neg      12      0          70     66      0     10      0      0      0   
4   neg   60874     na        1368    458      0      0      0      0      0   

   ...   ee_002  ee_003  ee_004  ee_005  ee_006  ee_007  ee_008 ee_009 ef_000  \
0  ...  1240520  493384  721044  469792  339156  157956   73224      0      0   
1  ...   421400  178064  293306  245416  133654   81140   97576   1500      0   
2  ...   277378  159812  423992  409564  320746  158022   95128    514      0   
3  ...      240      46      58      44      10       0       0      0      4   
4  ...   622012  229790  405298  347188  286954  311560  433954   1218      0   

  eg_000  
0      0  
1      0  
2      0  
3     32  
4      0  

[5 rows x 171 columns]

object    170
int64       1
dtype: int64


I have Identified the following: Object = 171, Int64 = 1 & dtype = int64

# Identfying Atributes

In summary, identifying attributes and observations is a critical step in data analysis and modeling. It forms the basis for understanding the dataset's structure, selecting relevant features, and ensuring data quality. This information enables informed decisions, helps with feature engineering, and ultimately leads to more accurate and effective data analysis and modeling. 

In [6]:
num_attributes = len(aps_failure_df.columns)
print("Number of Attributes:", num_attributes)

Number of Attributes: 171


In [7]:
num_observations = aps_failure_df.shape[0]
print("Number of Observations:", num_observations)

Number of Observations: 60000


# Identfying Dublicate dataset

In [ ]:
aps_failure_df.duplicated().sum()

Zero Dublication were found in this dataset.

# Identyfing Missing data: 

In [ ]:
aps_failure_df.isnull().sum()

•The class column contains the labels and its appears that the labels are being represented with the value of Zero’s

•Lengh equals to 171

•Data Types equales to int64

# Idenfying what these Characterisation mean.

The dataset i'm currently analysing consist of 600 rows (Observatios) & 171 Colums (Attributes), This makes it a modertely large dataset. Among these attributes, there are 171 distinc ones.

In my analysis, I'm taking steps to identify the duplicated data entries, which helps maintain the datasets intergrity. Addtionally, I will be addrressing the missing values  (Na, NAN) to ensure that any fields with non-numeric or null vales are appropriatley transformed into numerica data.

# 2. Exploratory Data Analysis

Exploratory data analysis I have found to be a curitical step in the data analysis process for serval important reasons are as understanding the dataset, this will allow me to explore thestructure, conects and charartics of the data. EDA process will hopefully allow me to identfy the patterns and relationships within my dataset, this will also allow myself to spot trends, anomalies and dependencies between variabl that might not be apparent at fist. 

Data Quality Assurance, will help me identify and address data quality issues such as missing values, outliers and inconsistencies. By addressing this early will enable myself to ensure the anaylsis is clean and reliable. Below as you can see I have deployed aps_failure_df.describe(include ="object") to identified the NA values.

In [ ]:
aps_failure_df.describe(include ="object")

As displayed above dataset, I have identified that there are dataset with NA in the column of ab_000, ad_000 which i have conclued that this would require to bechanged, although i still need to analys more of my dataset.

# Application of Data preparation & Evaluation methods 

# Cleaning Dataset Process 

This part of the process I will be working on identifying the unqiue value as well as the isnull values. As you can see that aps.failure_df has shown that there are quit a few datasets that are unique and now we can move to identifying isnull dataset to unsure that am working with consistant datasets.

In [ ]:
aps_failure_df["ab_000"].unique()

In [ ]:
aps_failure_df["ab_000"].isnull().sum()

As you can see above that there is 0 isnull with sum. am happy to continue with the cleaning of my dataset.

In [ ]:
aps_failure_df["ad_000"].unique()

In order to define the unique values within my dataset

In [ ]:
aps_failure_df["ad_000"].isnull().sum()

Next Step is to identfy the missing value using the missing_value as this will be valuable for the dataset analysis preprocessing. It is important to identfy the missing values in each columns (Attributes) within my dataset. The reason for this process as i would agree that its very much criticial as it helps me understand the qualitgy of dataset and also highlighting key area's that require cleaning or renaming dataset.

In [ ]:
missing_values=["NA","N/A", "-", "NaN", "missing", "na", "?"]

In order to identify and locate any missing value, i have decided to use the missing_value = ['na'] as part of my dataset cleaning process.

In [ ]:
aps_failure_df=pd.read_csv("aps_failure_set (2).csv" , na_values=missing_values)

My next step to chaning the missing values to number will be as followed.

In [ ]:
missing_value=aps_failure_df.isnull().sum()

In order to identify and quantify the missing vaules within APS_Failure_DF, I need this code to calculate the number of missing vailes for each column and store the results in the missing_value "variables".

In [ ]:
print(missing_value)

As you can see that the missing valued has now been replaced with unique values with the application of using EDA method, Although that been said, With the remaining columns that are no longer required can be removed. 
by doing this i have gained the valuable insight into the data quality which will allow me to take informed action.

It's key to understand which columns have missing data, especially how much data is missing in each column. this allows myself to take action that will be beneficial for my dataset and results.

# Removing - Droping Values - Changing Values

This Steps will disucss and highlight the process of removing the columns are as follow:

In [ ]:
import pandas as pd
aps_failure_df = aps_failure_df.replace('na', 0)

The reason why i have decided to us the import pandas as pd - asp_failure_df is to handle any missing or non-numerica value in the dataset am currently analysing, the aim of this code is to replace all instances of string 'na' with a numerica value of 0 in the entire dataframe.

By replacing the 'na' with a 0 this will make the data suitable for calculations, statisical analysis. which will allow the data to avoid any unecessary issues related to non-numeric or missing values in my analysis.

In [ ]:
aps_failure_df=aps_failure_df.drop(aps_failure_df.columns[aps_failure_df.isnull().mean()>10])

In [ ]:
aps_failure_df.fillna(aps_failure_df.median(),inplace=True)

In [ ]:
aps_failure_df.describe

# Re-naming 

Renaming my verables or features, I have decided not to take this option as the dataset i have am happy to identfy. 

# Clearned Dataset

In [ ]:
aps_failure_df.head(100)

# Exploratory Data Analysis - Visualisation

Exploratory Data Analysis or (EDA) is understanding the data set by summarizing its main characteristics and often plotting them visually. This step is very important especially when we arrive at modelling the data to apply Machine learning. Plotting in EDA consists of Histograms, Box plot, Scatter plots and many more.

Through the process of EDA, we can also refine the problem statement or definition of our problem.

# Boxplot Diagram

In [ ]:
%matplotlib inline
import pandas as pd
import seaborn as sns
import statistics as Stat
import matplotlib.pyplot as plt

In [ ]:
sns.boxplot(data=aps_failure_df)

Looking at diagram ca_000 to da_000

In [ ]:
sns.boxplot(x=aps_failure_df['ca_000'])

In [ ]:
sns.boxplot(x=aps_failure_df['da_000'])

Looking at diagram ee_000 to aa_000

In [ ]:
sns.boxplot(x=aps_failure_df['ee_000'])

In [ ]:
sns.boxplot(x=aps_failure_df['aa_000'])

# Scatter Chart for Random Range 

In [ ]:
N = 171
x = np.random.rand(N)
y = np.random.rand(N)
colors = np.random.rand(N)
area = (30 * np.random.rand(N))**2  # 0 to 15 point radii

plt.scatter(x, y, s=area, c=colors, alpha=0.5)
plt.show()

The purpose of this code is to create a scatter plot with some specific charaterticis. 
As you can see with the scatter plot allow the visualistion of the data point in two-dimenisional space. in this case x & y represent the coordinates of this point. 

I decided to use the ramndon data, to test the visualisation with random data.

The Scatter plot has generated 171 data points, where each point jas a random position (x,y) a ransome size. The scatter plot will clearly show these point are distrubuted in two-dimenal space. I believe its a useful tool for exploring data, checking paterns and unsertanding variables may relate to eachother.

In [ ]:
class_count = aps_failure_df["class"].value_counts()
print(class_count)

In [ ]:
class_counts = pd.Series([5900, 1000], index=['Negative', 'Positive'])
plt.figure(figsize=(6, 4))
class_counts.plot(kind= 'bar', color=['Blue', 'purple'])
plt.xlabel('Class')
plt.ylabel('Units')
plt.title('Air Pressur system-Apps Failure')
plt.xticks(range(len(class_count.index)), ['Negative', 'Positive'], rotation=0)
plt.legend(['The Nagative is not Related to Air Pressure system'])
plt.show()

In this sections i will be covering "what data are we exploring in this assignment, I will be looking

We are going to look at a data set on cars called “cardata.csv”.

The data contains more than 6,000 rows and more than 171 columns which have features

Assessing a Specific col

# Step 4: Principal Component Analysis

 Principal Component Analysis, In this section, i will be explore what is perhaps one of the most broadly used of unsupervised algorithms, principal component analysis (PCA). PCA is fundamentally a dimensionality reduction algorithm, but it can also be useful as a tool for visualization, for noise filtering,

# Number of Features Needed for Retain 99.5% Variance

In [ ]:
from sklearn.preprocessing import LabelEncoder
L_encoder = LabelEncoder()
aps_failure_df["class"]=L_encoder.fit_transform(aps_failure_df["class"])
aps_failure_df.head()

The code above am using scikit-learn library to encove categorical lables into nume3rical vales as this is my goal in specififcally, coverting the class column in the aps_failure_df. The main purpose is tp prepare the target veriables for machine learning algorithms. The labelEncoder will assigne a unique number to each unique class column making it easier ML model.

In [ ]:
from sklearn.decomposition import PCA

In [ ]:
aps_failure_df_no_label = aps_failure_df.drop(aps_failure_df["class"])
pca = PCA().fit(aps_failure_df_no_label) 
plt.plot(np.cumsum(pca.explained_variance_ratio_)) 
plt.xlabel('number of components')
plt.ylabel('cumulative explained variance')
plt.axhline(0.9950,color= "orangered",alpha=.5,ls="--")
plt.axvline(2,color= "orangered",alpha=.5,ls="--");
plt.xlim(0,10)

The main aim is to retain 99.5% variance data from the graphic show above, my understanding is that i must keep my features no higher or below 2 in order to get the results i am looking for.

The Graph above clearly show's that i would need it to be 2 Components to achive my objective of hitting a 99.5%

Analysing the implement PCA to dimensionally reduce the data to the number of features that you have discovered

In [ ]:
pca = PCA(2)
projected = pca.fit_transform(aps_failure_df_no_label)

In [ ]:
projected.shape

Side notes: observatins (Rows) equalling to 59998 and having 10 columns (Attributes)

# Identfying Array's within the dataset

In [ ]:
projected

The Array's format clearly displays that each of my obsersavtion has number of values of 2, although not the orginal values were present in the orginal dataset.

in order to understand the dataframe better and have a clear picture i have clearly displayed below: for the purpose of demonstrations, I have labled the colums all 10 of them below for visual purpose. 

# Identfying PCA Ptojected Columns

In [ ]:
failure_pca = pd.DataFrame(projected, columns =['C1','C2',])

In [ ]:
failure_pca.head()

As you can see that all my columns are labled with C1 to C10. 

In [ ]:
failure_pca["class"] = aps_failure_df["class"]

In [ ]:
failure_pca.head()

Data set I want to work with

As you can see that I will be selecting only the Frist Colum from C1 to C2. just before it get to the last column class. 

PCA was used to establish the minimum number of features needed for retaining 99.5% variance in the data and then implement PCA to dimensionally reduce the data to the number of features that you have discovered. 

This will Include a clear and concise explanation of what you i will be doing with the data 

and why you are doing it.

# Step 5:Curse of Dimensionality

I believe that curse of dimensionality referes to limited the issue of having limited obeservations, that being said having a resonable features. However, in order to identify the pattern within the dataset provided aps_failure_df, the options of requiring to removing the features which will allow the increase of observations which will lead the analysis a clear roadmap f the dataset. 

To achieve this, Princial component anlysis will play a crurial role, Although one must take in to consideration that one must implemente the cleaning stage before hand which reqires the Exploratory Data Analysis. With that process, this has highlighted the opertunity to remove the feature Class. This process is a must if the data would remain accurate dataset.I believe that this is a must process end steps in order to plan for appling michine learning. 


# Conclusion

in conclusion, I believe that the data preperation has been very crucial component of my analysis, laying down the foundation for my subsequent data drive assignment. Thoughout this assignment, I have been meticulous with cleaning the raw data to ensure its quality importing  from the library, The process involved serveral key steps, including loading data from the library, cleaning, handling missing data.

Continuing with cleaning and analysing the dataset, I had found that there had still columns with the catagory value of "na" & "nan" which i have identified as not being value which will cause some issues in the long run. 

The only solution for this dilema is to change the value with a "uniqu feature". 

In order to achieve this, I have implement the EDA process by indentifing all columns with more than >10% missing value would be dropped and replaced with the medium.

The nest steps i took was with the Exploratory Data Analysis - Visualisation, I have taken the optertunity to use multiple visualisation graphs which are listed below: 

1. (5) Boxplot: My boxplot clearly indicates that there are outliers on the top of the graphs, with random features.
      1.  [ca_000 to da_000]
      2.   [ee_000 to aa_000]
      
      
2. Scatter Char (1) Indicates the relationship between the different features. Athough in this case, I am looking in to random features. 


3. Histogram: (1) : The reason for histogram graph is to demostrate the differance units regarding the "Negative" & "Postive" values that are within the targetted veriables class. This process was crucial as it indicated an insight to what  was and what was not connected to the components failures for the ASP system.


Finally, the columns class was labelled from category value to numerical vaule in order to start the PCA process.


Implementing Princiaple Component Analysis not only assisted, has also highlighted the how many features that I would need to retain 99.5% of the cumulative explained variance of the dataset.

As you can clearly see in the graph that it shows in order to retain the 99.5% i would need to us (2) numbers of components with the array to lolcate the pattern from the dataset its self to reach my aimed result.

I honest Believe that these steps were crucial with supporting my analysis. 


# Reference

1.	Camp, D. (2023). Pandas Read csv() Tutorial: Importing Data. [online] https://www.datacamp.com/tutorial/pandas-read-csv. Available at: https://www.datacamp.com/tutorial/pandas-read-csv.

2.	Geeksforgeeks (2018a). Get unique values from a column in Pandas DataFrame. [online] GeeksforGeeks. Available at: https://www.geeksforgeeks.org/get-unique-values-from-a-column-in-pandas-dataframe/ [Accessed 10 Dec. 2018].

3.	Geeksforgeeks (2018b). Pandas isnull() and notnull() Method. [online] GeeksforGeeks. Available at: https://www.geeksforgeeks.org/python-pandas-isnull-and-notnull/?ref=gcse [Accessed 1 Nov. 2023].

4.	Idris, I. (2011). NumPy 1.5 : beginner’s guide. Birmingham: Packt Publishing.

5.	John Paul Mueller and Luca Massaron (2023). Python for Data Science For Dummies. John Wiley & Sons.

6.	Nik (2021). Pandas: Number of Rows in a Dataframe (6 Ways) • datagy. [online] datagy. Available at: https://datagy.io/pandas-number-of-rows/.

7.	Pandas (2023a). Chart Visualization — pandas 1.3.3 documentation. [online] pandas.pydata.org. Available at: https://pandas.pydata.org/pandas-docs/stable/user_guide/visualization.html.

8.	Pandas (2023b). pandas.DataFrame.describe — pandas 1.0.3 documentation. [online] pandas.pydata.org. Available at: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html.

9.	Pandas (2023c). Working with missing data — pandas 1.2.4 documentation. [online] pandas.pydata.org. Available at: https://pandas.pydata.org/pandas-docs/stable/user_guide/missing_data.html.

10.	Pratusevich, M. (2017a). Practice Python. [online] Practicepython.org. Available at: https://www.practicepython.org/.

11.	Pratusevich, M. (2017b). Practice Python. [online] Practicepython.org. Available at: https://www.practicepython.org/.

12.	Saravanan (2021). Import Pandas as pd. [online] Medium. Available at: https://medium.com/analytics-vidhya/import-pandas-as-pd-26344cd7235e.

13.	Scikit Learn (2009). sklearn.decomposition.PCA — scikit-learn 0.20.3 Documentation. [online] Scikit-learn.org. Available at: https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html [Accessed 2023].

14.	Seaborn (2023). seaborn.boxplot — seaborn 0.11.1 documentation. [online] seaborn.pydata.org. Available at: https://seaborn.pydata.org/generated/seaborn.boxplot.html.

15.	Stack Overflow (2014a). Is there a way to determine the order of labels in scikit-learn’s LabelEncoder? [online] Stack Overflow. Available at: https://stackoverflow.com/questions/25343411/is-there-a-way-to-determine-the-order-of-labels-in-scikit-learns-labelencoder [Accessed 1 Nov. 2023].

16.	Stack Overflow (2014b). Python - Principal Components Analysis Using Pandas Dataframe. [online] Stack Overflow. Available at: https://stackoverflow.com/questions/23282130/principal-components-analysis-using-pandas-dataframe.

17.	Stack Overflow (2018). understand df.isnull.mean() in python. [online] Stack Overflow. Available at: https://stackoverflow.com/questions/53283142/understand-df-isnull-mean-in-python.

18.	Stack Overflow (2019). How to run Python Code with ‘%matplotlib inline’? [online] Stack Overflow. Available at: https://stackoverflow.com/questions/55687832/how-to-run-python-code-with-matplotlib-inline.

19.	Stack Overflow (2021). How to Generate Random Colors in matplotlib? [online] Stack Overflow. Available at: https://stackoverflow.com/questions/14720331/how-to-generate-random-colors-in-matplotlib [Accessed 1 Nov. 2023].

20.	Stack overflow (2015). Python scikit learn pca.explained_variance_ratio_ cutoff. [online] Stack Overflow. Available at: https://stackoverflow.com/questions/32857029/python-scikit-learn-pca-explained-variance-ratio-cutoff.
21.	Stack overflow (2018). Understanding inplace=True in pandas. [online] Stack Overflow. Available at: https://stackoverflow.com/questions/43893457/understanding-inplace-true-in-pandas [Accessed 1 Nov. 2023].

22.	StackExchange (2019). What Does the PCA().transform() Method do? [online] Cross Validated. Available at: https://stats.stackexchange.com/questions/409176/what-does-the-pca-transform-method-do [Accessed 1 Nov. 2023].

23.	StackOverflow (2012). How to run Python Code with ‘%matplotlib inline’? [online] Stack Overflow. Available at: https://stackoverflow.com/questions/55687832/how-to-run-python-code-with-matplotlib-inline [Accessed 1 Nov. 2023].

24.	Stackoverflow (2011). r - Subset of rows containing NA (missing) values in a chosen column of a data frame. [online] Stack Overflow. Available at: https://stackoverflow.com/questions/7980622/subset-of-rows-containing-na-missing-values-in-a-chosen-column-of-a-data-frame.

25.	Stackoverflow (2016). How to resolve AttributeError: ‘DataFrame’ object has no attribute. [online] Stack Overflow. Available at: https://stackoverflow.com/questions/38134643/how-to-resolve-attributeerror-dataframe-object-has-no-attribute [Accessed 1 Nov. 2023].

26.	Stackoverflow (2021). Python - isnull().sum() vs isnull().count(). [online] Stack Overflow. Available at: https://stackoverflow.com/questions/60249807/python-isnull-sum-vs-isnull-count [Accessed 1 Nov. 2023].

27.	Vishal (2019). Python Basic Exercise for Beginners. [online] PYnative. Available at: https://pynative.com/python-basic-exercise-for-beginners/.

28.	w3school (2023). Pandas DataFrame head() Method. [online] www.w3schools.com. Available at: https://www.w3schools.com/python/pandas/ref_df_head.asp.
